2D Context get/putImageData cache copies cached image data twice

kkinnunen-apple opened the pull request:
#### 83d967c4920fe7d5eaeb957ddc5e2e236abce6a2
<pre>
2D Context get/putImageData cache copies cached image data twice
<a href="https://bugs.webkit.org/show_bug.cgi?id=264927">https://bugs.webkit.org/show_bug.cgi?id=264927</a>
<a href="https://rdar.apple.com/118497211">rdar://118497211</a>

Reviewed by Cameron McCormack.

Store the PixelBuffer that was used for putImageData,
avoid copying the the ImageData redundantly.

Instead of copying during caching, copy with premultiply.
This way we omit one memcpy for caching and one premultiply for the real
putImageData.

To simplify the implementation, removes the conservative approach
where the caching would kick in at 3rd putImageData. Just cache
the first.

* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::cacheImageDataIfPossible):
(WebCore::CanvasRenderingContext2DBase::takeCachedImageDataIfPossible const):
(WebCore::CanvasRenderingContext2DBase::putImageData):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:

Canonical link: <a href="https://commits.webkit.org/270975@main">https://commits.webkit.org/270975@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/19ba692a9ce38d4fea8b35cbbbebe30edac24f12

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26535 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5150 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27782 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28746 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24266 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26944 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6954 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2558 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24228 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26795 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3977 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22792 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3501 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3548 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23777 "Found 1 new test failure: platform/mac/media/encrypted-media/fps-generateRequest.html (failure)") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29229 "") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24191 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24175 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/29229 "") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3580 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1784 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/29229 "") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5016 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6477 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4054 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3908 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2470,20 +2472,19 @@ void CanvasRenderingContext2DBase::putImageData(ImageData& data, int dx, int dy,
-    if (cacheImageDataIfPossible(data, { dx, dy }, sourceRect))
-        options.add(DidDrawOption::PreserveCachedImageData);
-
+    if (!sourceRect.isEmpty()) {
```
Can we return early if this is empty? Or is there a need to call `didDraw` for some reason.

kkinnunen-apple replied:
I'll do that separately

kkinnunen-apple committed:
2D Context get/putImageData cache copies cached image data twice
https://bugs.webkit.org/show_bug.cgi?id=264927
rdar://118497211

Reviewed by Cameron McCormack.

Store the PixelBuffer that was used for putImageData,
avoid copying the the ImageData redundantly.

Instead of copying during caching, copy with premultiply.
This way we omit one memcpy for caching and one premultiply for the real
putImageData.

To simplify the implementation, removes the conservative approach
where the caching would kick in at 3rd putImageData. Just cache
the first.

* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::cacheImageDataIfPossible):
(WebCore::CanvasRenderingContext2DBase::takeCachedImageDataIfPossible const):
(WebCore::CanvasRenderingContext2DBase::putImageData):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:

Canonical link: https://commits.webkit.org/270975@main
